### ISSUE TITLE
conda install py_gnome?

### ISSUE THREAD

**Issue by rsignell-usgs** [NOAA-ORR-ERD/PyGnome#12]
I'm a bit confused.   I see an `environment.yml` file here to build a `py_gnome` custom environment:
https://github.com/NOAA-ORR-ERD/PyGnome/blob/master/py_gnome/environment.yml
which worked fine when I did 
```
conda env create environment.yml
```
but the resulting environment does not contain py_gnome.   Thus running one of the scripts in this environment results in 
```
---> 19 from gnome import scripting
     20 from gnome.basic_types import datetime_value_2d
     21 

ImportError: No module named gnome
```

I then looked for `py_gnome` (or something similar) on the NOAA-ORR-ERD channel at https://anaconda.org/NOAA-ORR-ERD/packages but did not find anything that looked suitable. 

I see a conda-recipe in https://github.com/NOAA-ORR-ERD/PyGnome/tree/master/py_gnome/py_gnome_conda_recipe but it looks old, and there is no build.bat file.   

Can I build py_gnome on windows with anaconda?

**Comment by rsignell-usgs** [comment]
@ocefpaf, have you done this?

**Comment by ChrisBarker-NOAA** [comment]
We currently aren't building a conda package for py_gnome. It's moving fast, so building from source is the best bet. Here are instructions:

https://github.com/NOAA-ORR-ERD/PyGnome/blob/master/InstallingWithAnaconda.rst

and yes, it does build no Windows.

Maybe it's time for a conda package, though.

PS: the gitHub version is a bit old -- stay tuned for a  update to the code.

**Comment by ocefpaf** [comment]
> @ocefpaf, have you done this?

For conda-recipes IOOS?  Nope.

> Maybe it's time for a conda package, though.

Chris do want to do that and we will get the recipe from you repo or should we do that in IOOS?

**Comment by rsignell-usgs** [comment]
I notice that @daf has a linux64 conda package here
https://anaconda.org/DAF/pygnome

**Comment by ocefpaf** [comment]
> version 0.1.1, 4 months old

Latest commit 895b944  happened on Oct 14, 2015 and did not touch the code.  After that all commits are >= 4 months ago.  So I guess @daf's package is up-to-date with the source code.

**Comment by daf** [comment]
Ah, this is slowly coming back to me.  The second call to `setup()` in [this code path](https://github.com/NOAA-ORR-ERD/PyGnome/blob/0ecb501da7e16895fafeee6f62c0202047eaa505/py_gnome/setup.py#L424-L436) broke my build too.

As we were using PyGnome with WebGnomeAPI, I ended up not using the PyGnome package I built and copied the PyGnome directory over into WebGnomeAPI's directory, then made a docker image out of the whole thing - that way it wasn't building PyGnome as an independent, standalone conda package.  This means my conda package likely is broken.

**Comment by ocefpaf** [comment]
Thanks @daf!  I am trying to sort the linux build and, if I make any progress,I will ping you.

**Comment by ChrisBarker-NOAA** [comment]
I'm not sure anything has changed with the build scripts, but we'll push an updated version of the code to gitHub today.

We're not allowed to do our actual development work on gitHub -- so it gets kind of behind.....

**Comment by ocefpaf** [comment]
> We're not allowed to do our actual development work on gitHub 

So sending PRs here is pretty much useless?

**Comment by ChrisBarker-NOAA** [comment]
no -- it's fine -- we do use gitHub to communicate with outside
contributors. So we can merge a PR on gitHub, then merge that back into our
internal repos.

Though  ideally you'd be looking at a more up-to-date version -- which is
coming soon!

-CHB



On Wed, Jan 13, 2016 at 9:49 AM, Filipe <notifications@github.com> wrote:

> We're not allowed to do our actual development work on gitHub
>
> So sending PRs here is pretty much useless?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NOAA-ORR-ERD/PyGnome/issues/12#issuecomment-171377737>
> .
>



-- 

Christopher Barker, Ph.D.
Oceanographer

Emergency Response Division
NOAA/NOS/OR&R            (206) 526-6959   voice
7600 Sand Point Way NE   (206) 526-6329   fax
Seattle, WA  98115       (206) 526-6317   main reception

Chris.Barker@noaa.gov

**Comment by ChrisBarker-NOAA** [comment]
It looks like @oblute is reviving this effort -- let's use this issue for discussion.

**Comment by oblute** [comment]
@ChrisBarker-NOAA is this package only compatible with python 2.7?

**Comment by ChrisBarker-NOAA** [comment]
yes, py2.7 only for now -- though we are actively working on porting to py3.


If you want to help, I can push that branch to this repo.

**Comment by oblute** [comment]
@ChrisBarker-NOAA i'm not sure how big of a help i would be, but would be happy to have a look. any idea how far you are in this process? conda-forge isn't compatible with 2.7

**Comment by ChrisBarker-NOAA** [comment]
well, yes and no.

conda-forge is no longer building py2 by default, but it can be turned on.

And everything py_gnome needs is in conda-forge already.

-CHB

**Comment by oblute** [comment]
@ChrisBarker-NOAA doesn't pygnome require https://github.com/NOAA-ORR-ERD/PyNUCOS and https://github.com/NOAA-ORR-ERD/OilLibrary?

**Comment by ChrisBarker-NOAA** [comment]
PyNUCOS is in conda-forge:

https://anaconda.org/conda-forge/pynucos

you're right about OilLIbrary -- that was under fast development, so hadn't built a package for it. But we probably should.

It's py2-3 compatible, so I should be able to do that.

Though it is now an optional dependency in the latest version of PyGNOME (maybe not up on gitHub yet)

The other issue is py_gd and libgd. They are on conda-forge for Linux and Mac, but not the only Windows version is on our channel:

https://anaconda.org/search?q=py_gd

That's due to the challenge of a Windows build of ligd:

https://github.com/conda-forge/libgd-feedstock/issues/9

Still waiting for someone with Windows Building skills for that one (we hacked it together in a way not suitable for conda-forge)

So I guess there's a bit more work to do -- now you know why we haven't done it yet :-)

**Comment by oblute** [comment]
@ChrisBarker-NOAA if we were just considering releasing py3 functionality on unix, what are the steps remaining toward getting there?

**Comment by ChrisBarker-NOAA** [comment]
I"ve been working on a version that will run on both python2 and 3 -- I taking the philosophy of "python3 code that can run on python2" it's in the "py23" branch in this repo.

It currently will compile and run on python3, with 48 failing tests (out of 1200 or so, so not too bad :-) )

The most challenging part is an issue with a newer version of Cython -- but that may not be an issue for your use case, depending on what features you are using.

The latest version is in the "py23" branch.

**Comment by ChrisBarker-NOAA** [comment]
Keeping this open, as we should do it.

There are no major obstacles to making a conda package -- we just haven't done it -- still moving fast.

One issue though -- py_gnome won't build with pip -- pip tries to make an sdist first, and then built thaqt into a wheel. But the way we have the code structured (the C++ stuff), it doesn't make a proper sdist.

so `setup.py install--single-version-externally-managed`

is probably the way to go.

**Comment by ChrisBarker-NOAA** [comment]
There's a not-fully functional feedstock on conda-forge now:

https://github.com/conda-forge/pygnome-feedstock

We're having issues with segfaults in the test environment in conda-build -- aarrgghh! 

https://github.com/conda-forge/pygnome-feedstock/pull/5

hopefully resolved soon?
